lenskit-core/src/main/java/org/lenskit/data/dao/file/DelimitedColumnEntityFormat.java

```diff
@@ -0,0 +1,235 @@
+    }
+
+    /**
+     * Set the delimiter to use for this entity format.
+     * @param delim The delimiter to use.
+     */
+    public void setDelimiter(String delim) {
+        delimiter = delim;
+    }
+
+    /**
+     * Set whether to read a header.
+     * @param header `true` to read a header line from the file.
+     */
+    public void setHeader(boolean header) {
+        readHeader = header;
+        headerLines = 1;
+    }
+
+    /**
+     * Get the number of header lines to read.
+     * @return The number of header lines to read.
+     */
+    public int getHeaderLines() {
+        return headerLines;
+    }
+
+    /**
+     * Set the number of header lines to read.  Setting this **disables** {@link #setHeader(boolean)}.
+     * @param lines The number of header lines to read.
+     */
+    public void setHeaderLines(int lines) {
+        headerLines = lines;
+        readHeader = false;
+    }
+
+    /**
+     * Query whether this format uses the header line(s).
+     * @return `true` if the reader will parse a header line.
+     */
+    boolean usesHeader() {
+        return readHeader;
+    }
+
+    /**
+     * Set the entity type.
+     * @param type The entity type.
+     */
+    public void setEntityType(EntityType type) {
+        entityType = type;
+    }
+
+    /**
+     * Get the entity type.
+     * @return The entity type.
+     */
+    @Override
+    public EntityType getEntityType() {
+        return entityType;
+    }
+
+    /**
+     * Set the entity builder class.
+     * @param builder The entity builder class.
+     */
+    public void setEntityBuilder(Class<? extends EntityBuilder> builder) {
+        entityBuilder = builder;
+    }
+
+    /**
+     * Get the entity builder class.
+     * @return The entity builder class.
+     */
+    public Class<? extends EntityBuilder> getEntityBuilder() {
+        return entityBuilder;
+    }
+
+    /**
+     * Instantiate a new entity builder.
+     * @return A new entity builder.
+     */
+    public EntityBuilder newEntityBuilder() {
+        if (entityBuilderCtor == null || !entityBuilderCtor.getDeclaringClass().equals(entityBuilder)) {
+            try {
+                entityBuilderCtor = entityBuilder.getConstructor(EntityType.class);
+            } catch (NoSuchMethodException e) {
+                throw new IllegalArgumentException("cannot find suitable constructor for " + entityBuilder);
+            }
+        }
+        try {
+            return entityBuilderCtor.newInstance(entityType);
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new RuntimeException("could not instantiate entity builder", e);
+        }
+    }
+
+    /**
+     * Add a column.
+     * @param attr The attribute to add as a column, or `null` to skip the next column.
+     * @throws IllegalStateException if non-labeled columns have already been specified.
+     */
+    public void addColumn(@Nullable TypedName<?> attr) {
```

Comment:
`Label the next column. The first time this is called it specifies the label for the first column. The second time this is called it specifies the second column label and so forth` or something like that to make the ordering clear.